docs/developing-contracts/gear-program.md

```diff
@@ -5,25 +5,263 @@ sidebar_position: 2
+        let counter = self.votes_received.entry(name).or_insert(0);
+        *counter += 1;
+    }
+}
+
+// The state itself (i.e. the variable state will be accessed through)
+static mut STATE: State = State::new();
+```
+
+We will also need to define the metadata structure for implementing the input/output communication interface. The method described is a binary map for mutual interaction between different programming languages. For example, since the program is compiled into WASM format, it only understands the byte language. To simplify the operation, we define data structures in advance for further encoding and decoding. For this, we use a special macro gstd::metadata!:
+
+```rust
+gstd::metadata! {
+   title: "Voting App",
+   handle:
+       input: Action,
+   state:
+       input: StateAction,
+       output: StateReply,
+}
+```
+
+Now let's start processing incoming messages. Every time our contract receives an incoming message, we will process it accordingly. Let's describe the handle() function:
+
+```rust
+#[derive(Debug, TypeInfo, Encode)]
+pub enum StateReply {
+   All(BTreeMap<String, i32>),
+   VotesFor(i32),
+}
+ 
+#[derive(Debug, TypeInfo, Decode)]
+pub enum StateAction {
+   All,
+   VotesFor(String),
+}
+
+// Handle function that processes the incoming message
 #[no_mangle]
 pub unsafe extern "C" fn handle() {
-    msg::reply(b"Hello world!", 0, 0);
+    let action: Action = msg::load().unwrap();
+
+    debug!("Received action: {:?}", action);
+
+    match action {
+        Action::AddCandidate(name) => {
+            STATE.add_candidate(name.clone());
+
+            msg::reply((), 0, 0);
+
+            debug!("Added new candidate: {:?}", name);
+        }
+
+        Action::VoteForCandidate(name) => {
+            STATE.vote_for_candidate(name.clone());
+
+            msg::reply((), 0, 0);
+
+            debug!("Voted for: {:?}", name);
+        }
+    }
 }
+```
+
+Now we can communicate with our program. Add and vote for candidates. It remains to teach our program to show all candidates or votes for a certain one by name. To do this, we will use the meta_state() function, which will return the state immediately and without any cost.
 
+```rust
+// The function that returns a part of memory with a state
 #[no_mangle]
-pub unsafe extern "C" fn init() {}
+pub unsafe extern "C" fn meta_state() -> *mut [i32; 2] {
+   let query: StateAction = msg::load().expect("failed to decode input argument");
+ 
+   let encoded = match query {
+       StateAction::All => StateReply::All(STATE.votes_received.clone()).encode(),
+ 
+       StateAction::VotesFor(name) => {
+           let votes_for_candidate = STATE
+               .votes_received
+               .get(&name)
+               .expect("Can't find any candidate");
+ 
+           StateReply::VotesFor(votes_for_candidate.clone()).encode()
+       }
+   };
+ 
+   let result = gstd::macros::util::to_wasm_ptr(&encoded[..]);
+   core::mem::forget(encoded);
+ 
+   result
```

Comment:
It is recommended to use a new function for getting a pointer.
https://github.com/gear-tech/gear/blob/master/examples/meta/src/lib.rs#L167
```suggestion
   gstd::util::to_leak_ptr(encoded)
```